src/store/ParsingBuffer.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_STORE_PARSINGBUFFER_H
+#define SQUID_SRC_STORE_PARSINGBUFFER_H
+
+#include "sbuf/SBuf.h"
+#include "StoreIOBuffer.h"
+
+#include <optional>
+
+namespace Store
+{
+
+/// A continuous buffer for efficient accumulation and NUL-termination of
+/// Store-read bytes. The buffer accumulates two kinds of Store readers:
+///
+/// * Readers that do not have any external buffer to worry about but need to
+///   accumulate, terminate, and/or consume buffered content read by Store.
+///   These readers use the default constructor and then allocate the initial
+///   buffer space for their first read (if any).
+///
+/// * Readers that supply their StoreIOBuffer at construction time. That buffer
+///   is enough to handle the majority of use cases. However, the supplied
+///   StoreIOBuffer capacity may be exceeded when parsing requires accumulating
+///   multiple Store read results and/or NUL-termination of a full buffer.
+///
+/// This buffer seamlessly grows as needed, reducing memory over-allocation and,
+/// in case of StoreIOBuffer-seeded construction, memory copies.
+class ParsingBuffer
```

Comment:
This looks like a highly over-engineered. What we actually need is for StoreIOBuffer to be backed by a MemBlob instead of a raw memory allocation. Then it can spawn SBuf as-needed.

Reply:
> What we actually need is for StoreIOBuffer to be backed by a MemBlob instead of a raw memory allocation

StoreIOBuffer is used for too many very different/contradicting purposes, from a read-only content area, to an empty buffer, to an offset variable. We should not make it much smarter or more powerful because doing so will go against the needs of many of its users. We can and should replace StoreIOBuffer with multiple classes, each serving a particular use case, but this PR is not really about that (I actually _removed_ one of such replacement classes in branch commit 2ca3a3c to reduce this PR footprint -- we will come back to finish that work eventually).

> This looks like a highly over-engineered.

This ParsingBuffer class solves a real problem that cannot be solved with a StoreIOBuffer, SBuf, or MemBlob alone. As the code improves, the problem itself may disappear or transform, but, for now, the class is pretty much required if we want to avoid adding memory copies on the primary response processing path. The class itself is not very complex and cannot be made much simpler AFAICT. In fact, we did try to split it into two classes (based on two main caller categories), one of which would be a bit simpler, but the result was overall worse than what we have now.